apps/src/lab2/redux/predictLevelRedux.ts

```diff
@@ -0,0 +1,62 @@
+import {sendPredictLevelReport} from '@cdo/apps/code-studio/progressRedux';
+import {PayloadAction, createSelector, createSlice} from '@reduxjs/toolkit';
+import {setLoadedPredictResponse} from '@cdo/apps/lab2/lab2Redux';
+import {RootState} from '@cdo/apps/types/redux';
+
+export interface PredictLevelState {
+  response: string;
+  hasSubmittedResponse: boolean;
+}
+
+const initialState: PredictLevelState = {
+  // User's response for the level, if the level is a predict level. It is an empty string if this
+  // is not a predict level or if the user has not yet submitted a response.
+  response: '',
+  // If the user has submitted a predict response for the current level.
+  hasSubmittedResponse: false,
+};
+
+// SELECTORS
+
+// The predict answer is locked if the level does not allow multiple predict attempts
+// and the user has already submitted a response.
+export const isPredictAnswerLocked = createSelector(
+  [
+    (state: RootState) =>
+      state.lab.levelProperties?.predictSettings?.allowMultipleAttempts,
+    (state: RootState) => state.predictLevel.hasSubmittedResponse,
+  ],
+  (allowMultipleAttempts, hasSubmittedResponse) => {
+    return !allowMultipleAttempts && hasSubmittedResponse;
+  }
+);
+
+// REDUCER
+const predictSlice = createSlice({
+  name: 'predictLevel',
+  initialState,
+  reducers: {
+    setPredictResponse(state, action: PayloadAction<string>) {
+      state.response = action.payload;
+    },
+    setHasSubmittedPredictResponse(state, action: PayloadAction<boolean>) {
+      state.hasSubmittedResponse = action.payload;
+    },
+  },
+  extraReducers: builder => {
+    builder.addCase(sendPredictLevelReport.fulfilled, state => {
+      state.hasSubmittedResponse = true;
+    });
+    builder.addCase(setLoadedPredictResponse, (state, action) => {
+      // We have loaded a predict response from the server.
+      // If the response is not empty, we consider the user to have submitted a response.
+      state.response = action.payload;
+      state.hasSubmittedResponse = !!action.payload;
```

Comment:
Do we have validation on submitting something like a series of spaces somewhere? Could students lock themselves out of a predict level by submitting empty text, and would we want to protect against that? (Maybe we don't want to.)

Reply:
In our old version of predict levels, a student can submit a bunch of spaces and it will lock the answer. So for starters I think we can reproduce that behavior, but see if we want to improve on it as a follow up.